### PR TITLE
ODC-6778: Add console developer catalog customization

### PR DESCRIPTION
**Pull request by debsmita1**
**Story:**
https://issues.redhat.com/browse/ODC-6778

Add dev catalog types customization API based on https://github.com/openshift/enhancements/pull/1206

### CODE REVIEW

File: operator/v1/types_console.go
Comment (jerolimov) [inline operator/v1/types_console.go:117]: nit: its types or just types, or?

```suggestion
	// developerCatalog allows to configure the shown developer catalog categories (filters) and types (providers).
```

File: operator/v1/0000_70_console-operator.crd.yaml
Comment (jerolimov) [inline operator/v1/0000_70_console-operator.crd.yaml:129]: There are two spaces behind the type: string, so that the verify step doesn't pass:

```suggestion
                            type: string
```

File: operator/v1/types_console.go
Comment (jerolimov) [inline operator/v1/types_console.go:155]: The first word references the YAML/JSON type anywhere else. Recommend to align the doc to Quickstarts and add actions. See https://github.com/openshift/api/blob/874dd9fe45490eb7347db7422582e43ddffa2e3d/operator/v1/types_console.go#L192-L196 and https://github.com/openshift/api/blob/874dd9fe45490eb7347db7422582e43ddffa2e3d/operator/v1/types_console.go#L200-L206

```suggestion
	// disableTypes is a list of developer catalog IDs that are not shown to users.
	// +kubebuilder:validation:Optional
	// +kubebuilder:validation:MinItems=1
	// +optional
	DisableTypes []string `json:"disableTypes,omitempty"`
```

**Review by jerolimov — CHANGES_REQUESTED** [review]

**Comment by jerolimov** [comment]
/retitle ODC-6778: Add console developer catalog customization

File: operator/v1/types_console.go
Comment (jerolimov) [inline operator/v1/types_console.go:154]: If I'm correct we agreed on @JoelSpeed idea to add an `enabledTypes` as well, or?

We can maybe also add a hint where to find the types = sub-catalog IDs. Wdyt?

```suggestion
	// enabledTypes is a list of developer catalog types (sub-catalogs IDs) that will be shown to users.
	// Sub-catalogs are added via console plugins, the available types (sub-catalog IDs) are available
	// in the console on the cluster configuration page, or when editing the YAML in the console.
	// +optional
	EnabledTypes []string `json:"enabledTypes,omitempty"`
	// disabledTypes is a list of developer catalog types (sub-catalogs IDs) that are not shown to users.
	// Sub-catalogs are added via console plugins, the available types (sub-catalog IDs) are available
	// in the console on the cluster configuration page, or when editing the YAML in the console.
	// +optional
	DisabledTypes []string `json:"disabledTypes,omitempty"`
```

**Review by jerolimov — CHANGES_REQUESTED** [review]
Just one small comment, but in general this is fine:

File: operator/v1/types_console.go
Comment (jerolimov) [reply operator/v1/types_console.go:154]: > obvious to an end user?

Hopefully, at least the term is used in our UI and in our frontend code. Fyi:

image

**Review by jerolimov — APPROVED** [review]

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:158]: With enabled, there's a way to enable all (this is the default when neither list is configured). We then have a way to enable only certain items, and to disable only certain items.

There's currently no way to permanently disable all items (new catalogs could be added but not in the list).

I wonder if we need a wildcard option, let me see how others tackle this

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:169]: You need to add a comment that explains what happens when this config is omitted, IMO that should be all catalogs are shown (which would be backwards compatible)

File: operator/v1/types_console.go
Comment (deads2k) [reply operator/v1/types_console.go:146]: > Having reviewed other enabled/disabled pairs, it's an error to have the value in both. I think we should maintain that behaviour here. Do you have a webhook that validates this type already?

We promoted CEL to beta for CRD validation in 1.25.  If there isn't a webhook, I'm open to considering an IOU for CEL validation before we ship 4.12.

File: operator/v1/types_console.go
Comment (deads2k) [reply operator/v1/types_console.go:158]: > I wonder if we need a wildcard option, let me see how others tackle this

Interesting observation.  Kube admission configuration is often used as an example that allows a default list with default on and default off as an option.  Then configuration values that forceEnable and forceDisable.  Very much like (exactly like?) this API.  But we didn't consider the case of, "disable everything".  I'm open to a wildcard value like "*".

Comment (JoelSpeed) [reply operator/v1/types_console.go:158]: There is an alternative suggestion to this API as well, which is to make a discriminated union which would look something like below:
```
state: All|Enabled|Disabled|None
enabled: []
disabled: []
```

Though I'm concerned about the 4 values of the enum, I'm not sure it's intuitive that All and Enabled are different from one another, likewise with Disabled and None.

Perhaps if the list was All, Explicit and None and then the structure would be
```
state: All|Explicit|None
explicit:
  enabled: []
  disabled: []
```
I think I prefer the current approach to the DU approach personally

Comment (jerolimov) [reply operator/v1/types_console.go:158]: We're fine with both, please let us know which direction we should follow...

**Review by JoelSpeed — COMMENTED** [review]
Two observations of things to work on here:
- I think we should go with the enabled and disabled lists as they are right now, but I think we should add a special `*` value to allow users to disable all catalogs. Which will then need a comment in the disabled option to explain that
- Is there any way that a user can use a command to look up the valid values for these lists? If not, its my understanding they must look at the console, can we include in the docs a couple of examples of valid values that would be added to these lists

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:160]: Can we try
```suggestion
	// +kubebuilder:validation:UniqueItems:=true
	Disabled []string `json:"disabled"`
```

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:154]: Can we try
```suggestion
	// +kubebuilder:validation:UniqueItems:=true
	Disabled []string `json:"disabled"`
```

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:147]: We either need a webhook or CEL (https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules) for validating that an item is in either enabled or disabled but not both

The CEL validation would be for checking two listSets are disjoint
```
self.set1.all(e, !(e in self.set2))
```
I'll work out how this is applied as this is new for K8s 1.25

File: operator/v1/types_console.go
Comment (JoelSpeed) [reply operator/v1/types_console.go:158]: In this discriminated union:
* Default state to Enabled for backwards compatibility 
* Enabled + empty enabled[]: show everything
* Enabled + non-empty enabled[]: show only the items in the enabled list
* Disabled + empty disabled[]: disable everything
* Disabled + non-empty disabled[]: disable only the items in the disabled list
```
state: Enabled|Disabled
enabled: []
disabled: []
```
This means that you cannot use enabled and disabled at the same time, which, as we don't have specific lists of enabled and disabled by default (default is all on), we don't need to use them at the same time

File: operator/v1/types_console.go
Comment (debsmita1) [reply operator/v1/types_console.go:147]: we don't need this if we add state as a unionDiscriminator

File: operator/v1/types_console.go
Comment (debsmita1) [reply operator/v1/types_console.go]: used `+listType=set`

File: operator/v1/types_console.go
Comment (JoelSpeed) [inline operator/v1/types_console.go:156]: Based on the testing I'm doing here you may need a couple of changes to make sure this works. You have a required field within an optional struct, we need to make sure it's omitted correctly.

https://github.com/openshift/api/blob/b4c38862d26f227af4221d01dd41331e2b78201a/example/v1/types_stable.go#L70-L87

I think you need `omitempty` for state, and then we can add the CEL as in that example for the enabled and disabled fields here

**Comment by JoelSpeed** [comment]
Have you done any manual testing of the API to check that when you don't set the developer console new config, that it doesn't complain about a required field? Just want to check the validations here

**Comment by debsmita1** [comment]
@JoelSpeed This is what I have tested
https://user-images.githubusercontent.com/22490998/191085477-7acc8daa-f3ba-4d7b-ad31-edd66cb67b80.mov

**Comment by JoelSpeed** [comment]
Thanks for the video, are you happy to wait for the rebase to 1.25 to land before you merge this? I think it should happen this week. Once 1.25 is merged we can test this properly

The alternative is to build off of #1298 and add some tests for this project based on what we are doing there (eg #1297 for CPMS), that could help prove the validations

**Comment by jerolimov** [comment]
Tested this API via https://github.com/openshift/console-operator/pull/676 on a cluster bot.

Could verify that all the API change works fine.

/label qe-approved
/lgtm

**Comment by jerolimov** [comment]
Propagate from the ODC Epic:
/label px-approved
/label docs-approved

**Comment by JoelSpeed** [comment]
I'm happy with this API and that testing will follow up once the 1.25 rebase has landed, deferring to @deads2k for final approval

File: authorization/v1/0000_03_authorization-openshift_01_rolebindingrestriction.crd.yaml
Comment (deads2k) [inline authorization/v1/0000_03_authorization-openshift_01_rolebindingrestriction.crd.yaml:214]: why was this removed?

**Comment by deads2k** [comment]
verify passed so I'll assume the change in rolebindingrestrictions was correct.

/approve

**Comment by JoelSpeed** [comment]
/hold

I disagree, I don't think the served and storage versions can be removed, pretty sure if we try to install that CRD into a cluster it will not allow the CRD to be created

**Comment by JoelSpeed** [comment]
Yep, just tried to add an integration test for this and the API returned errors trying to install the CRD
```
                              {
                                  Type: "FieldValueInvalid",
                                  Message: "Invalid value: []apiextensions.CustomResourceDefinitionVersion{apiextensions.CustomResourceDefinitionVersion{Name:\"v1\", Served:false, Storage:false, Deprecated:false, DeprecationWarning:(*string)(nil), Schema:(*apiextensions.CustomResourceValidation)(nil), Subresources:(*apiextensions.CustomResourceSubresources)(nil), AdditionalPrinterColumns:[]apiextensions.CustomResourceColumnDefinition(nil)}}: must have exactly one version marked as storage version",
                                  Field: "spec.versions",
                              },
                              {
                                  Type: "FieldValueInvalid",
                                  Message: "Invalid value: []string(nil): must have at least one stored version",
                                  Field: ...
```

**Comment by JoelSpeed** [comment]
This [commit](https://github.com/JoelSpeed/api/commit/fc37c0532e03bdb6a5c46b641a2ed313860645fc) can help demonstrate that the CRD change here is invalid, if you pull this branch (based on this PR) and from the root of the API repo run `make -C authorization/v1 test` it will error right now

File: operator/v1/types_console.go
Comment (jerolimov) [inline operator/v1/types_console.go:262]: Nit: A generic type name `ResourceAttributesAccessReview` with a really specific job "defines the visibility of the perspective", `PerspectiveVisibilityAccessReview` might be an option - or updating the doc.

**Comment by jerolimov** [comment]
> I disagree, I don't think the served and storage versions can be removed, pretty sure if we try to install that CRD into a cluster it will not allow the CRD to be created

Thanks for finding and blocking this. It's fixed now. As this was the last blocker we can continue with this. We still can add some tests when your testing framework is ready.

/unhold
/lgtm

**Pull request merged**
